modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,305 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.Lifecycle;
+import org.elasticsearch.common.scheduler.SchedulerEngine;
+import org.elasticsearch.common.scheduler.TimeValueSchedule;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.List;
+import java.util.function.LongSupplier;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * This service will implement the needed actions (e.g. rollover, retention) to manage the data streams with a DLM lifecycle configured.
+ * It runs on the master node and it schedules a job according to the configured {@link DataLifecycleService#DLM_POLL_INTERVAL_SETTING}.
+ */
+public class DataLifecycleService implements ClusterStateListener, Closeable, SchedulerEngine.Listener {
+
+    public static final String DLM_POLL_INTERVAL = "indices.dlm.poll_interval";
+    public static final Setting<TimeValue> DLM_POLL_INTERVAL_SETTING = Setting.timeSetting(
+        DLM_POLL_INTERVAL,
+        TimeValue.timeValueMinutes(10),
+        TimeValue.timeValueSeconds(1),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private static final Logger logger = LogManager.getLogger(DataLifecycleService.class);
+    /**
+     * Name constant for the job DLM schedules
+     */
+    private static final String DATA_LIFECYCLE_JOB_NAME = "dlm";
+
+    private final Settings settings;
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ResultDeduplicator<TransportRequest, Void> transportActionsDeduplicator;
+    private final LongSupplier nowSupplier;
+    private final Clock clock;
+    private volatile boolean isMaster = false;
+    private volatile TimeValue pollInterval;
+    private SchedulerEngine.Job scheduledJob;
+    private final SetOnce<SchedulerEngine> scheduler = new SetOnce<>();
+
+    public DataLifecycleService(
+        Settings settings,
+        Client client,
+        ClusterService clusterService,
+        Clock clock,
+        ThreadPool threadPool,
+        LongSupplier nowSupplier
+    ) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.clock = clock;
+        this.transportActionsDeduplicator = new ResultDeduplicator<>(threadPool.getThreadContext());
+        this.nowSupplier = nowSupplier;
+        this.scheduledJob = null;
+        this.pollInterval = DLM_POLL_INTERVAL_SETTING.get(settings);
+        clusterService.addListener(this);
```

Comment:
Reminder for later: We can add the listener after the constructor has finished initializing because now we are leaking `this` before initialization has finished.

Reply:
Thanks, great point ! Added it